ident/map.go

```diff
@@ -0,0 +1,256 @@
+		return BinaryID(checked.NewBytes(bytes, nil))
+	}
+	return m.opts.Pool.Clone(key)
+}
+
+// Get returns a value in the map for an identifier if found.
+func (m *Map) Get(id ID) (ValueType, bool) {
+	hash := m.hash(id)
+	for v, ok := m.lookup[hash]; ok; v, ok = m.lookup[hash] {
+		if v.key.id.Equal(id) {
+			return v.value, true
+		}
+		// Linear probe to "next" to this entry (really a rehash)
+		hash++
+	}
+	var empty ValueType
+	return empty, false
+}
+
+// Set will set the value for an identifier.
+func (m *Map) Set(id ID, value ValueType) {
+	m.set(id, value, mapSetOptions{
+		copyKey:     true,
+		finalizeKey: true,
+	})
+}
+
+// SetNoCopyNoFinalizeKey will set the value for an identifier but will
+// avoid copying the identifier and will not finalize the key if/when evicted.
+func (m *Map) SetNoCopyNoFinalizeKey(id ID, value ValueType) {
+	m.set(id, value, mapSetOptions{
+		copyKey:     false,
+		finalizeKey: false,
+	})
+}
+
+type mapSetOptions struct {
+	copyKey     bool
+	finalizeKey bool
+}
+
+func (m *Map) set(id ID, value ValueType, opts mapSetOptions) {
+	hash := m.hash(id)
+	for v, ok := m.lookup[hash]; ok; v, ok = m.lookup[hash] {
+		if v.key.id.Equal(id) {
+			m.lookup[hash] = MapEntry{
+				key:   v.key,
+				value: value,
+			}
+			return
+		}
+		// Linear probe to "next" to this entry (really a rehash)
+		hash++
+	}
+
+	var keyID ID
+	if opts.copyKey {
+		keyID = m.copyKey(id)
+	} else {
+		keyID = id
+	}
+
+	key := mapKey{
+		id:       keyID,
+		finalize: opts.finalizeKey,
+	}
+
+	m.lookup[hash] = MapEntry{
+		key:   key,
+		value: value,
+	}
+}
+
+// Iter provides the underlying map to allow for using a native Go for loop
+// to iterate the map, however callers should only ever read and not write
+// the map.
+func (m *Map) Iter() map[MapHash]MapEntry {
+	return m.lookup
+}
+
+// Len returns the number of map entries in the map.
+func (m *Map) Len() int {
+	return len(m.lookup)
+}
+
+// Has returns true if value exists for key, false otherwise, it is
+// shorthand for a call to Get that doesn't return the value.
+func (m *Map) Has(id ID) bool {
+	_, ok := m.Get(id)
+	return ok
+}
+
+// Delete will remove a value set in the map for the specified key.
+func (m *Map) Delete(id ID) {
+	hash := m.hash(id)
+	for v, ok := m.lookup[hash]; ok; v, ok = m.lookup[hash] {
+		if v.key.id.Equal(id) {
+			delete(m.lookup, hash)
+			if v.key.finalize {
+				v.key.id.Finalize()
+			}
+			return
+		}
+		// Linear probe to "next" to this entry (really a rehash)
+		hash++
+	}
+}
+
+// Reset will reset the map by simply deleting all keys to avoid
+// allocating a new map.
+func (m *Map) Reset() {
+	for hash, v := range m.lookup {
+		delete(m.lookup, hash)
+		if v.key.finalize {
+			v.key.id.Finalize()
+		}
+	}
+}
+
+// ResetByReallocateMap will avoid deleting all keys and reallocate
+// a new map, this is useful if you believe you have a larage map and
+// will not need to grow back to a similar size.
+func (m *Map) ResetByReallocateMap() {
```

Comment:
Is this really needed? Feels like users could just call the ctor if they need to call this, and this is probably not going to be called that often so the overhead of one extra pointer is probably a worthy tradeoff for smaller API surface area.

Reply:
They could call the constructor again yes, however I did want it to make it clear to the caller on what they were doing when they did Reset and give them an alternative that's very simple to avoid deleting all the keys if they really wanted to.